index/compaction.go

```diff
@@ -0,0 +1,319 @@
+package index
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/parquet-go/parquet-go"
+
+	"github.com/polarsignals/frostdb/dynparquet"
+	"github.com/polarsignals/frostdb/parts"
+)
+
+const (
+	IndexFileExtension     = ".idx"
+	ParquetCompactionTXKey = "compaction_tx"
+	dirPerms               = os.FileMode(0o755)
+	filePerms              = os.FileMode(0o640)
+)
+
+type Compaction func(w io.Writer, compact []parts.Part, options ...parquet.WriterOption) (int64, error)
+
+type FileCompaction struct {
+	// settings
+	dir     string
+	compact Compaction
+	maxSize int64
+
+	// internal data
+	indexFiles []*os.File
+	offset     int64 // Writing offsets into the file
+	ref        int64 // Number of references to file.
+
+	// Options
+	logger log.Logger
+}
+
+func NewFileCompaction(dir string, maxSize int64, compact Compaction, logger log.Logger) (*FileCompaction, error) {
+	f := &FileCompaction{
+		dir:     dir,
+		compact: compact,
+		maxSize: maxSize,
+		logger:  logger,
+	}
+
+	if err := os.MkdirAll(dir, dirPerms); err != nil {
+		return nil, err
+	}
+
+	return f, nil
+}
+
+func (f *FileCompaction) MaxSize() int64 { return f.maxSize }
+
+// Snapshot will create a new index file and all future compacitons are written to that index file.
+func (f *FileCompaction) Snapshot(dir string) error {
```

Comment:
Can there be some concurrency issues in cases where we snapshot this file while we're writing to it?

Reply:
No because we call this from the LSM Snapshot function which ensures that no compactions are running.